File: LLWallPaper.App/Utils/AppPaths.cs
using System.IO;
using System.Linq;

namespace LLWallPaper.App.Utils;

public static class AppPaths
{
    public static string Root
    {
        get
        {
            var basePath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            return Path.Combine(basePath, "AkaakuHub", "LLWallPaper");
        }
    }

    public static string SettingsPath => Path.Combine(Root, "settings.json");
    public static string FavoritesPath => Path.Combine(Root, "favorites.json");
    public static string HistoryPath => Path.Combine(Root, "history.json");
    public static string LegacyHistoryPath => Path.Combine(Root, "history.jsonl");
    public static string LogPath => Path.Combine(Root, "logs", "app.log");
    public static string CacheRoot => Path.Combine(Root, "cache", "images");

    public static string GetCachePathForKey(string key)
    {
        if (string.IsNullOrWhiteSpace(key))
        {
            return string.Empty;
        }

        var safeId = string.Concat(key.Select(ch => char.IsLetterOrDigit(ch) ? ch : '_'));
        var fileName = $"card_{safeId}_full.webp";
        return Path.Combine(CacheRoot, fileName);
    }

    public static void EnsureDirectories()
    {
        Directory.CreateDirectory(Root);
        Directory.CreateDirectory(Path.GetDirectoryName(LogPath)!);
        Directory.CreateDirectory(CacheRoot);
    }
}

File: LLWallPaper.Mac/Sources/LLWallPaperMacCore/Infrastructure/AppPaths.swift
import Foundation

public enum AppPaths {
  public static var root: URL {
    let baseUrl = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[
      0]
    return baseUrl.appending(path: "LLWallPaper", directoryHint: .isDirectory)
  }

  public static var settingsUrl: URL {
    root.appending(path: "settings.json")
  }

  public static var favoritesUrl: URL {
    root.appending(path: "favorites.json")
  }

  public static var historyUrl: URL {
    root.appending(path: "history.json")
  }

  public static var logUrl: URL {
    root.appending(path: "logs/app.log")
  }

  public static var cacheRoot: URL {
    root.appending(path: "cache/images", directoryHint: .isDirectory)
  }

  public static func cacheUrl(for key: String) -> URL? {
    guard !key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      return nil
    }

    let safeId = key.map { character in
      character.isLetter || character.isNumber ? String(character) : "_"
    }.joined()
    return cacheRoot.appending(path: "card_\(safeId)_full.webp")
  }

  public static func ensureDirectories() throws {
    try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    try FileManager.default.createDirectory(
      at: logUrl.deletingLastPathComponent(),
      withIntermediateDirectories: true
    )
    try FileManager.default.createDirectory(at: cacheRoot, withIntermediateDirectories: true)
  }
}
